{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: PandasAI in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (1.3.3)\n",
      "Requirement already satisfied: astor<0.9.0,>=0.8.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (0.8.1)\n",
      "Requirement already satisfied: duckdb<0.9.0,>=0.8.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (0.8.1)\n",
      "Requirement already satisfied: ipython<9.0.0,>=8.13.1 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from PandasAI) (8.13.2)\n",
      "Requirement already satisfied: matplotlib<4.0.0,>=3.7.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (3.7.1)\n",
      "Requirement already satisfied: openai<0.28.0,>=0.27.5 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (0.27.10)\n",
      "Requirement already satisfied: pandas==1.5.3 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (1.5.3)\n",
      "Requirement already satisfied: pydantic<2,>=1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (1.10.12)\n",
      "Requirement already satisfied: python-dotenv<2.0.0,>=1.0.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (1.0.0)\n",
      "Requirement already satisfied: scipy<2.0.0,>=1.9.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (1.10.1)\n",
      "Requirement already satisfied: sqlalchemy<2.0.0,>=1.4.49 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from PandasAI) (1.4.49)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from pandas==1.5.3->PandasAI) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from pandas==1.5.3->PandasAI) (2023.3)\n",
      "Requirement already satisfied: numpy>=1.21.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from pandas==1.5.3->PandasAI) (1.24.3)\n",
      "Requirement already satisfied: backcall in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.2.0)\n",
      "Requirement already satisfied: decorator in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.18.2)\n",
      "Requirement already satisfied: matplotlib-inline in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.1.6)\n",
      "Requirement already satisfied: pickleshare in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.7.5)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (3.0.38)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (2.15.1)\n",
      "Requirement already satisfied: stack-data in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.6.2)\n",
      "Requirement already satisfied: traitlets>=5 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (5.9.0)\n",
      "Requirement already satisfied: pexpect>4.3 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (4.8.0)\n",
      "Requirement already satisfied: appnope in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from ipython<9.0.0,>=8.13.1->PandasAI) (0.1.3)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (1.0.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (4.39.4)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (1.4.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (23.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (9.5.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from matplotlib<4.0.0,>=3.7.1->PandasAI) (3.0.9)\n",
      "Requirement already satisfied: requests>=2.20 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from openai<0.28.0,>=0.27.5->PandasAI) (2.31.0)\n",
      "Requirement already satisfied: tqdm in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from openai<0.28.0,>=0.27.5->PandasAI) (4.66.1)\n",
      "Requirement already satisfied: aiohttp in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from openai<0.28.0,>=0.27.5->PandasAI) (3.8.5)\n",
      "Requirement already satisfied: typing-extensions>=4.2.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from pydantic<2,>=1->PandasAI) (4.5.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from jedi>=0.16->ipython<9.0.0,>=8.13.1->PandasAI) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from pexpect>4.3->ipython<9.0.0,>=8.13.1->PandasAI) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython<9.0.0,>=8.13.1->PandasAI) (0.2.6)\n",
      "Requirement already satisfied: six>=1.5 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from python-dateutil>=2.8.1->pandas==1.5.3->PandasAI) (1.16.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests>=2.20->openai<0.28.0,>=0.27.5->PandasAI) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests>=2.20->openai<0.28.0,>=0.27.5->PandasAI) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests>=2.20->openai<0.28.0,>=0.27.5->PandasAI) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests>=2.20->openai<0.28.0,>=0.27.5->PandasAI) (2023.5.7)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (23.1.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (6.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (4.0.3)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (1.9.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (1.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp->openai<0.28.0,>=0.27.5->PandasAI) (1.3.1)\n",
      "Requirement already satisfied: executing>=1.2.0 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from stack-data->ipython<9.0.0,>=8.13.1->PandasAI) (1.2.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from stack-data->ipython<9.0.0,>=8.13.1->PandasAI) (2.2.1)\n",
      "Requirement already satisfied: pure-eval in /Users/mikeredshaw/Library/Python/3.11/lib/python/site-packages (from stack-data->ipython<9.0.0,>=8.13.1->PandasAI) (0.2.2)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install PandasAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "import pandas as pd\n",
    "import pandas as pd\n",
    "from pandasai import PandasAI\n",
    "from pandasai.llm.openai import OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup OpenAI API\n",
    "OPENAI_API_KEY = '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'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"/Users/mikeredshaw/Documents/Schulich MBAN/Deloitte ACP/Data Sets/Sample - Superstore.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandasai/__init__.py:145: UserWarning: `PandasAI` (class) is deprecated since v1.0 and will be removed in a future release. Please use `SmartDataframe` instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "llm = OpenAI(api_token=OPENAI_API_KEY)\n",
    "pandas_ai = PandasAI(llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[Errno 30] Read-only file system: '/exports'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m/mredshaw/DS2/GPT_Data_2.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://github/mredshaw/DS2/GPT_Data_2.ipynb#X11sdnNjb2RlLXZmcw%3D%3D?line=0'>1</a>\u001b[0m pandas_ai\u001b[39m.\u001b[39;49mrun(df, prompt\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mHow many columns are there in the dataset?\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandasai/__init__.py:199\u001b[0m, in \u001b[0;36mPandasAI.run\u001b[0;34m(self, data_frame, prompt, show_code, anonymize_df, use_error_correction_framework)\u001b[0m\n\u001b[1;32m    196\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(data_frame, \u001b[39mlist\u001b[39m):\n\u001b[1;32m    197\u001b[0m     data_frame \u001b[39m=\u001b[39m [data_frame]\n\u001b[0;32m--> 199\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dl \u001b[39m=\u001b[39m SmartDatalake(data_frame, config)\n\u001b[1;32m    200\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dl\u001b[39m.\u001b[39mchat(prompt)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandasai/smart_datalake/__init__.py:77\u001b[0m, in \u001b[0;36mSmartDatalake.__init__\u001b[0;34m(self, dfs, config, logger, memory, cache)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\n\u001b[1;32m     63\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m     64\u001b[0m     dfs: List[Union[DataFrameType, Any]],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     68\u001b[0m     cache: Cache \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m     69\u001b[0m ):\n\u001b[1;32m     70\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[39m    Args:\u001b[39;00m\n\u001b[1;32m     72\u001b[0m \u001b[39m        dfs (List[Union[DataFrameType, Any]]): List of dataframes to be used\u001b[39;00m\n\u001b[1;32m     73\u001b[0m \u001b[39m        config (Union[Config, dict], optional): Config to be used. Defaults to None.\u001b[39;00m\n\u001b[1;32m     74\u001b[0m \u001b[39m        logger (Logger, optional): Logger to be used. Defaults to None.\u001b[39;00m\n\u001b[1;32m     75\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 77\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49minitialize()\n\u001b[1;32m     79\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_load_config(config)\n\u001b[1;32m     81\u001b[0m     \u001b[39mif\u001b[39;00m logger:\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandasai/smart_datalake/__init__.py:121\u001b[0m, in \u001b[0;36mSmartDatalake.initialize\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    119\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mValueError\u001b[39;00m:\n\u001b[1;32m    120\u001b[0m     charts_dir \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mjoin(os\u001b[39m.\u001b[39mgetcwd(), \u001b[39m\"\u001b[39m\u001b[39mexports\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mcharts\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 121\u001b[0m os\u001b[39m.\u001b[39;49mmakedirs(charts_dir, mode\u001b[39m=\u001b[39;49m\u001b[39m0o777\u001b[39;49m, exist_ok\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[1;32m    123\u001b[0m \u001b[39m# Create /cache folder if it doesn't exist\u001b[39;00m\n\u001b[1;32m    124\u001b[0m \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[0;32m<frozen os>:215\u001b[0m, in \u001b[0;36mmakedirs\u001b[0;34m(name, mode, exist_ok)\u001b[0m\n",
      "File \u001b[0;32m<frozen os>:225\u001b[0m, in \u001b[0;36mmakedirs\u001b[0;34m(name, mode, exist_ok)\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 30] Read-only file system: '/exports'"
     ]
    }
   ],
   "source": [
    "pandas_ai.run(df, prompt=\"How many columns are there in the dataset?\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
